### PR TITLE
Replace np.einsum with np.tensordot in _upsampled_dft

### PR DESCRIPTION
**Pull request by taylor-scott**
## Description

In feature/register_translation.py, _upsampled_dft uses np.einsum to compute the upsampled DFT of an image. This is not easily generalizable and in numpy 1.11 is (extremely) slow because no optimize flag is available.

np.einsum can be replaced with sequential calls to np.tensordot which (should) generalize to higher dimensions.

E.g.,

    import numpy as np
    image = np.random.uniform(10, size=(10, 15, 20))
    kernels = [np.random.uniform(10, size=(2*i, i)) for i in image.shape]
    einsum_result = np.einsum('ijk,li,mj,nk->lmn', image, *kernels, optimize=True)
    tensordot_result = np.copy(image)
    for kernel in kernels[::-1]:
        tensordot_result = np.tensordot(kernel, tensordot_result, axes=(1, -1))
    
    print(np.abs(tensordot_result - einsum_result).max()) # 0.0

It should also be faster in numpy 1.11 and, from testing on my latptop, not slower than np.einsum in numpy 1.16. It seems to be slightly faster for some inputs. 

Is there a standard way to benchmark functions? I see there's a benchmarks folder but I'm not sure how to use it.

## Todo
- [x] Benchmark in numpy 1.16 (current) ~and numpy 1.11~
- [x] Extend to nD images
- [x] Write tests for nD images
- [x] Clean up kernel statement

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] ~Gallery example in `./doc/examples` (new features only)~
- [x] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by jni** [comment]
@taylor-scott very nice! I'm not super concerned with supporting NumPy 1.11, which is getting a bit long in the tooth, but the code is significantly cleaner now so I think this is a worthwhile change either way!

Regarding benchmarks, I think it's a good idea to add a couple (one for 2D and one for 3D), since performance is part of the motivation for this PR. I recommend you look at https://github.com/scikit-image/scikit-image/blob/8135c2f3505d77134162846abf1b06c0b18d5a56/benchmarks/benchmark_morphology.py for how to parameterise tests, and then add a `class RegisterTranslation` to `benchmark_feature.py`. As long as the class has a `setup` method and a `time_register_translation` method, it'll get picked up by airspeed velocity for benchmarks.

**Comment by jni** [comment]
By the way, if this work sounds too onerous, we can do it the old-fashioned way: run a benchmark script on your machine and share with us the script and the results. =)

**Comment by hmaarrfk** [comment]
Honestly, I wouldn't worry about numpy 1.11 much. It is almost 3 years old now, so not so important.

I should write these out somewhere:

```
pip install asv
```
or 
```
conda install asv
```

Then to test your benchmark
```
asv run -E existing -b RankSuite
```

To compare your results to master
```
asv continuous -b RankSuite -E conda:3.7 master simplify_benchmarks 
```

you might have to swap the order the branch names.

File: skimage/feature/register_translation.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def _upsampled_dft(data, upsampled_region_size,
 
     Parameters
     ----------
-    data : 2D or 3D ndarray
+    data : nD ndarray
```
Comment (stefanv) [inline skimage/feature/register_translation.py]: Just `ndarray`?

**Comment by stefanv** [comment]
On Mon, 04 Feb 2019 21:24:22 -0800, Taylor D. Scott wrote:
> Thoughts on using tensordot in the 2D case? It's slower, but the code
> cleaner with no special cases.

I think 50 µsec is nothing to be concerned about at all, especially
given that the code is clear and expressive.

**Comment by taylor-scott** [comment]
I think I'm nearly done with this... 

A couple more things I noticed:

1. `_upsampled_dft` calculates the DFT but we actually want the inverse DFT. That's way the function call is `_upsampled_df(image_product.conj(), ...).conj()`. 

I think a good way to handle this is to add an optional flag `inverse=False` to `_uspampled_dft`. If true, use positive `im2pi` when calculating the kernel (to give the inverse DFT). If false, use `-im2pi` as it is now. 

The `conj` calls can get expensive for large images (~10 msec for a 1040x1392) so I think it's good to get rid of them if possible. 

2. [L256](https://github.com/scikit-image/scikit-image/blob/a3d2557209fb5913609749f06e70b367cef00617/skimage/feature/register_translation.py#L256): This might be returning the right thing by coincidence but max() sorts lexicographically for complex numbers. Looking at the matlab code, it's supposed to be the maximum cross_correlation sorted by absolute value. I'll fix this one tonight.

**Comment by stefanv** [comment]
On Tue, 05 Feb 2019 18:04:00 +0000, Taylor D. Scott wrote:
> 1. `_upsampled_dft` calculates the DFT but we actually want the inverse DFT. That's way the function call is `_upsampled_df(image_product.conj(), ...).conj()`. 
> 
> I think a good way to handle this is to add an optional flag
> `inverse=False` to `_uspampled_dft`. If true, use positive `im2pi`
> when calculating the kernel (to give the inverse DFT). If false, use
> `-im2pi` as it is now.

Could you expand a bit more on this item?

> 2. [L256](https://github.com/scikit-image/scikit-image/blob/a3d2557209fb5913609749f06e70b367cef00617/skimage/feature/register_translation.py#L256):
> This might be returning the right thing by coincidence but max() sorts
> lexicographically for complex numbers. Looking at the matlab code,
> it's supposed to be the maximum cross_correlation sorted by absolute
> value. I'll fix this one tonight.

Thanks!

**Comment by taylor-scott** [comment]
The [MATLAB code ](https://www.mathworks.com/matlabcentral/fileexchange/18401-efficient-subpixel-image-registration-by-cross-correlation) is here:

    CCabs = abs(CC);
    [rloc, cloc] = find(CCabs == max(CCabs(:)),1,'first');
    CCmax = CC(rloc,cloc);

It takes the absolute value of the cross-correlation, finds the argmax, then return the cross-correlation at that point.

The code in register_translation, `CCmax = cross_correlation.max()` returns the maximum value based on how numpy sorts complex numbers, which is not by absolute value. Instead, it sorts by real part first, then by complex part.

So:

    np.array([1+1j, 0+10j]).max()

gives `1 + 1j` even though `abs(0+10j) > abs(1 + 1j)`

See the discussion here: numpy/numpy#8151

**Comment by stefanv** [comment]
@taylor-scott Sorry if my response was confusing; I was asking about item #1.

**Comment by taylor-scott** [comment]
@stefanv I'm sorry, I completely misread your question.

To find the cross-correlation, we take the inverse DFT of the image product. In pixel-level registration, this is just `cross_correlation = ifftn(image_product)` but for subpixel registration the function calls `_upsampled_dft` which embeds the image product into a large image and calculates the DFT of the larger image.

This almost gives the cross-correlation, but we actually need the inverse DFT of the larger image. So to calculate the cross-correlation, `register_translation` uses the [identity ](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Expressing_the_inverse_DFT_in_terms_of_the_DFT) `fft(image.conj).conj/N = ifft(image)` where `N` is the image size.

For large images, the time to call `conj` is not trivial. For example, calling `image.conj()` on one of my images (1040 x 1392) takes ~11 msec and registering two images takes ~230 msec total. 

To avoid this call to `conj`, I proposed adding an optional flag `inverse=False` to `_upsampled_dft`. When false, return the DFT as the function does now. When true, return the inverse DFT which is what we want in this case.

This is easy to accomplish while generating the kernel:
```python
    kernel = ((np.arange(upsampled_region_size[0]) - ax_offset)[:, None]
              * np.fft.fftfreq(n_items, upsample_factor))
    if inverse:
        # We can also add the normalizing factor here if desired
        kernel = np.exp(im2pi * kernel)
    else:
        kernel = np.exp(-im2pi * kernel)
```

**Comment by taylor-scott** [comment]
Also, after thinking about it a bit more, this may be an issue better handled in a separate pull request after some more discussion. I think `_upsampled_dft` could be generally useful, but IMO this aspect and how the image is embedded in the upsampled image are non-intuitive.

**Comment by stefanv** [comment]
@taylor-scott Thank you for that very clear explanation.  Yikes, such subtle tricks are very easy to miss, and I agree with you that making it explicit would be an improvement.  If you are willing to make the change in another PR, then great—let's get this one merged so long.  Feel free to add your name to CONTRIBUTORS.txt if you want.

@scikit-image/core Ready for final review.

**Comment by taylor-scott** [comment]
Ok, great. I've removed the inverse flag from this PR and I'll open a separate PR to address that. Thanks @stefanv !

**Comment by stefanv** [comment]
Thanks, @taylor-scott!

**Pull request merged**